Signal/src/ViewControllers/OWSSoundSettingsViewController.m

```diff
@@ -0,0 +1,165 @@
+
+- (void)viewDidAppear:(BOOL)animated
+{
+    [self updateTableContents];
+}
+
+- (void)updateNavigationItems
+{
+    self.navigationItem.leftBarButtonItem =
+        [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel
+                                                      target:self
+                                                      action:@selector(cancelWasPressed:)];
+
+    if (self.isDirty) {
+        self.navigationItem.rightBarButtonItem =
+            [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemSave
+                                                          target:self
+                                                          action:@selector(saveWasPressed:)];
+    } else {
+        self.navigationItem.rightBarButtonItem = nil;
+    }
+}
+
+#pragma mark - Table Contents
+
+- (void)updateTableContents
+{
+    OWSTableContents *contents = [OWSTableContents new];
+
+    __weak OWSSoundSettingsViewController *weakSelf = self;
+
+    OWSTableSection *soundsSection = [OWSTableSection new];
+    soundsSection.headerTitle = NSLocalizedString(
+        @"NOTIFICATIONS_SECTION_SOUNDS", @"Label for settings UI that allows user to change the notification sound.");
+    
+    NSArray<NSNumber *> *allSounds;
+    switch (self.soundType) {
+        case OWSSoundType_Notification:
+            allSounds = [OWSSounds allNotificationSounds];
+            break;
+        case OWSSoundType_Ringtone:
+            allSounds = [OWSSounds allRingtoneSounds];
+            break;
+    }
+    for (NSNumber *nsValue in allSounds) {
+        OWSSound sound = (OWSSound)nsValue.intValue;
+        OWSTableItem *item;
+        if (sound == self.currentSound) {
+            item = [OWSTableItem checkmarkItemWithText:[OWSSounds displayNameForSound:sound]
+                                           actionBlock:^{
+                                               [weakSelf soundWasSelected:sound];
+                                           }];
+        } else {
+            item = [OWSTableItem actionItemWithText:[OWSSounds displayNameForSound:sound]
+                                        actionBlock:^{
+                                            [weakSelf soundWasSelected:sound];
+                                        }];
+        }
+        [soundsSection addItem:item];
+    }
+
+    [contents addSection:soundsSection];
+
+    self.contents = contents;
+}
+
+#pragma mark - Events
+
+- (void)soundWasSelected:(OWSSound)sound
+{
+    [self.audioPlayer stop];
+    self.audioPlayer = [OWSSounds audioPlayerForSound:sound];
+    // Suppress looping in this view.
+    self.audioPlayer.numberOfLoops = 0;
+    [self.audioPlayer play];
+
+    if (self.currentSound == sound) {
+        return;
+    }
+
+    self.currentSound = sound;
+    self.isDirty = YES;
+    [self updateTableContents];
+    [self updateNavigationItems];
+}
+
+- (void)cancelWasPressed:(id)sender
+{
+    // TODO: Add "discard changes?" alert.
```

Comment:
I didn't bother with a warning when abandoning unsaved changes, but I did include a dirty flag towards this.

Reply:
👍